core/src/main/scala/org/bitcoins/core/crypto/MnemonicCode.scala

```diff
@@ -0,0 +1,257 @@
+
+    entropyNoChecksum ++ checksum == entropyWithChecksum
+
+  }
+
+  def words: Vector[String]
+
+  /**
+    * Returns the entropy initially provided to construct
+    * this mnemonic code
+    */
+  def toEntropy: BitVector = {
+    val entropyWithChecksumBits = toEntropyWithChecksum
+    val lengthNoEntropy = MnemonicCode
+      .getMnemonicCodeInfo(words)
+      .entropyBits
+
+    entropyWithChecksumBits.take(lengthNoEntropy)
+  }
+
+  /**
+    * Returns the entropy _with checksum_ originally provided
+    * to construct this mnemonic code
+    */
+  def toEntropyWithChecksum: BitVector = {
+
+    def bitsForWord(word: String): BitVector = {
+
+      // tail-recursive binary search for word in word list
+      // stolen from https://codereview.stackexchange.com/questions/161887/re-inventing-the-wheel-binary-search-in-scala
+      import MnemonicCode.{ENGLISH_WORDS => wordlist}
+      @tailrec
+      def search(
+          start: Int = 0,
+          end: Int = wordlist.length - 1
+      ): Option[Int] = {
+        val mid = start + (end - start) / 2
+        if (start > end) None // can't be found
+        else if (wordlist(mid) == word) Some(mid) // found
+        else if (wordlist(mid) > word)
+          search(start, mid - 1) // narrow the field
+        else search(mid + 1, end)
+      }
+
+      val indexOpt = search()
+
+      val index = indexOpt.getOrElse(
+        throw new IllegalArgumentException(s"Couldn't find $word in word list!")
+      )
+
+      BitVector.fromInt(index, size = MnemonicCode.BIT_GROUP_LENGTH)
+
+    }
+
+    val bits: BitVector = words
+      .map(curr => bitsForWord(curr))
+      .fold(BitVector.empty)((accum, curr) => accum ++ curr)
+
+    val codeInfo = MnemonicCode.getMnemonicCodeInfo(words)
+
+    bits.take(codeInfo.entropyAndChecksumBits)
+  }
+}
+
+/**
+  * @see [[https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki BIP39]]
+  */
+object MnemonicCode {
+
+  private[crypto] val BIT_GROUP_LENGTH = 11
+
+  private case class MnemonicCodeImpl(words: Vector[String])
+      extends MnemonicCode
+
+  /**
+    * Creates a mnemonic code from the provided words
+    */
+  def fromWords(words: Vector[String]): MnemonicCode = MnemonicCodeImpl(words)
+
+  /**
+    * Generates a mnemonic code from the given entropy bytes
+    */
+  def fromEntropy(entropy: ByteVector): MnemonicCode =
+    fromEntropy(entropy.bits)
+
+  /**
+    * The minimum number of entropy bits needed to construct a mnemonic code
+    */
+  private[crypto] val MIN_ENTROPY_BITS = 128
+
+  /**
+    * The maximum number of entropy bits allowed to construct mnemonic code
+    */
+  private[crypto] val MAX_ENTROPY_BITS = 256
+
+  /**
+    * The entropy for a mnemonic code has to be a multiple of 32
+    */
+  private[crypto] val ENTROPY_MULTIPLE = 32
+
+  /**
+    * Generates a mnemonic code from the given entropy bits
+    */
+  def fromEntropy(entropy: BitVector): MnemonicCode = {
+    require(
```

Comment:
Can't we move this invariants into the abstract class?

Reply:
These `require` statements are checking the validity of the provided arguments to the function, though. I'd say it makes more sense to have them in the function. Fail fast, and all that. Open to changing it